---
## title: "Lesson 3: Changing your data"
output: html_document
---

## Background for this activity
In this activity, you’ll review a scenario, and focus on manipulating and changing real data in R. You will learn more about functions you can use to manipulate your data, use statistical summaries to explore your data, and gain initial insights for your stakeholders. 

Throughout this activity, you will also have the opportunity to practice writing your own code by making changes to the code chunks yourself. If you encounter an error or get stuck, you can always check the Lesson3_Change_Solutions .rmd file in the Solutions folder under Week 3 for the complete, correct code. 

## The Scenario
In this scenario, you are a junior data analyst working for a hotel booking company. You have been asked to clean a .csv file that was created after querying a database to combine two different tables from different hotels. You have already performed some basic cleaning functions on this data; this activity will focus on using functions to conduct basic data manipulation.

## Step 1: Load packages

Start by installing the required packages. If you have already installed and loaded `tidyverse`, `skimr`, and `janitor` in this session, feel free to skip the code chunks in this step. This may take a few minutes to run, and you may get a pop-up window asking if you want to proceed. Click yes to continue installing the packages. 

```{r install packages}
install.packages("tidyverse")
install.packages("skimr")
install.packages("janitor")
```

Once a package is installed, you can load it by running the `library()` function with the package name inside the parentheses:

```{r load packages}
library(tidyverse)
library(skimr)
library(janitor)
```

## Step 2: Import data

In the chunk below, you will use the `read_csv()` function to import data from a .csv in the project folder called "hotel_bookings.csv" and save it as a data frame called `hotel_bookings`. 

If this line causes an error, copy in the line setwd("projects/Course 7/Week 3") before it. 

Type the file name in the correct place to read it into your R console: 

```{r load dataset}
hotel_bookings <- read_csv("")
```

## Step 3: Getting to know your data

Like you have been doing in other examples, you are going to use summary functions to get to know your data. This time, you are going to complete the code chunks below in order to use these different functions. You can use the `head()` function to preview the columns and the first several rows of data. Finish the code chunk below and run it:

```{r head function}
head()
```

In [1]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('hotel_bookings.csv')

df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

### Practice Quiz 

1. How many columns are in this dataset?
A: 45
B: 100
C: 32
D: 60

2. The 'arrival_date_month' variable is chr or character type data.  
A: True
B: False

Now you know this dataset contains information on hotel bookings. Each booking is a row in the dataset, and each column contains information such as what type of hotel was booked, when the booking took place, and how far in advance the booking took place (the 'lead_time' column).

In addition to `head()` you can also use the `str()` and `glimpse()` functions to get summaries of each column in your data arranged horizontally. You can try these two functions by completing and running the code chunks below:

```{r str function}
(hotel_bookings)
```

You can see the different column names and some sample values to the right of the colon. 

```{r glimpse function}
glimpse()
```

You can also use `colnames()` to get the names of the columns in your dataset. Run the code chunk below to get the column names:

```{r colnames function}
colnames(hotel_bookings)
```


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## Manipulating your data

Let's say you want to arrange the data by most lead time to least lead time because you want to focus on bookings that were made far in advance. You decide you want to try using the `arrange()` function; input the correct column name after the comma and run this code chunk: 

```{r arrange function}
arrange(hotel_bookings, )
```


But why are there so many zeroes? That's because `arrange()` automatically orders by ascending order, and you need to specifically tell it when to order by descending order, like the below code chunk below:

```{r arrange function descending} 
arrange(hotel_bookings, desc(lead_time))
```

Now it is in the order you needed. You can click on the different pages of results to see additional rows of data, too.  

## Practice Quiz
What is the highest lead time for a hotel booking in this dataset?
A: 737
B: 709
C: 629
D: 0

In [11]:
df.sort_values(by='lead_time', ascending=False)

hotel  is_canceled  lead_time  arrival_date_year  \
1      Resort Hotel            0        737               2015   
4182   Resort Hotel            0        709               2016   
65240    City Hotel            1        629               2017   
65249    City Hotel            1        629               2017   
65245    City Hotel            1        629               2017   
...             ...          ...        ...                ...   
85725    City Hotel            0          0               2016   
85726    City Hotel            0          0               2016   
85758    City Hotel            0          0               2016   
85759    City Hotel            0          0               2016   
20017  Resort Hotel            0          0               2016   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
1                   July                        27                          1   
4182            February                         9                         25   
65240              March                        13                         30   
65249              March                        13                         30   
65245              March                        13                         30   
...                  ...                       ...                        ...   
85725              March                        13                         20   
85726              March                        13                         20   
85758              March                        13                         20   
85759              March                        13                         20   
20017            January                         3                         12   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
1                            0                     0       2  ...   
4182                         8                    20       2  ...   
65240                        0                     2       2  ...   
65249                        0                     2       2  ...   
65245                        0                     2       2  ...   
...                        ...                   ...     ...  ...   
85725                        1                     0       2  ...   
85726                        1                     0       1  ...   
85758                        1                     0       2  ...   
85759                        1                     0       2  ...   
20017                        0                     1       2  ...   

       deposit_type  agent company days_in_waiting_list    customer_type  \
1        No Deposit    NaN     NaN                    0        Transient   
4182     No Deposit    NaN     NaN                    0        Transient   
65240    Non Refund    1.0     NaN                    0        Transient   
65249    Non Refund    1.0     NaN                    0        Transient   
65245    Non Refund    1.0     NaN                    0        Transient   
...             ...    ...     ...                  ...              ...   
85725    No Deposit    9.0     NaN                    0        Transient   
85726    No Deposit    NaN     NaN                    0        Transient   
85758    No Deposit    NaN     NaN                    0        Transient   
85759    No Deposit    NaN     NaN                    0  Transient-Party   
20017    No Deposit    NaN   250.0                    0        Transient   

         adr  required_car_parking_spaces  total_of_special_requests  \
1        0.0                            0                          0   
4182    68.0                            0                          0   
65240   62.0                            0                          0   
65249   62.0                            0                          0   
65245   62.0                            0                          0   
...      ...                          ...                        ...   
85

Notice that when you just run `arrange()` without saving your data to a new data frame, it does not alter the existing data frame. Check it out by running `head()` again to find out if the highest lead times are first: 

```{r head function part two}
head(hotel_bookings)
```

This will be true of all the functions you will be using in this activity. If you wanted to create a new data frame that had those changes saved, you would use the assignment operator, <- , as written in the code chunk below to store the arranged data in a data frame named 'hotel_bookings_v2':

```{r new dataframe}
hotel_bookings_v2 <-
  arrange(hotel_bookings, desc(lead_time))
```

Run `head()`to check it out: 

```{r new dataframe part two}
head(hotel_bookings_v2)
```

You can also find out the maximum and minimum lead times without sorting the whole dataset using the `arrange()` function. Try it out using the max() and min() functions below:

```{r}
max(hotel_bookings$lead_time)
```

```{r}
min(hotel_bookings$lead_time)
```

Remember, in this case, you need to specify which dataset and which column using the $ symbol between their names. Try running the below to see what happens if you forget one of those pieces:

```{r}
min(lead_time)
```

This is a common error that R users encounter. To correct this code chunk, you will need to add the data frame and the dollar sign in the appropriate places. 

Now, let's say you just want to know what the average lead time for booking is because your boss asks you how early you should run promotions for hotel rooms. You can use the `mean()` function to answer that question since the average of a set of number is also the mean of the set of numbers:

```{r mean}
mean(hotel_bookings$lead_time)
```

You should get the same answer even if you use the v2 dataset that included the `arrange()` function. This is because the `arrange()` function doesn't change the values in the dataset; it just re-arranges them.

```{r mean part two}
mean(hotel_bookings_v2$lead_time)
```

## Practice Quiz 
What is the average lead time?
A: 100.0011
B: 45.0283
C: 14.0221
D: 104.0114

In [14]:
df['lead_time'].mean()

104.01141636652986

You were able to report to your boss what the average lead time before booking is, but now they want to know what the average lead time before booking is for just city hotels. They want to focus the promotion they're running by targeting major cities. 

You know that your first step will be creating a new dataset that only contains data about city hotels. You can do that using the `filter()` function, and name your new data frame 'hotel_bookings_city':

```{r filter}
 <- 
  filter(hotel_bookings, hotel_bookings$hotel=="City Hotel")
```

Check out your new dataset:

```{r new dataset}
head(hotel_bookings_city)
```

You quickly check what the average lead time for this set of hotels is, just like you did for all of hotels before:

```{r average lead time city hotels}
mean(hotel_bookings_city$lead_time)
```

Now, your boss wants to know a lot more information about city hotels, including the maximum and minimum lead time. They are also interested in how they are different from resort hotels. You don't want to run each line of code over and over again, so you decide to use the `group_by()`and`summarize()` functions. You can also use the pipe operator to make your code easier to follow. You will store the new dataset in a data frame named 'hotel_summary':

```{r group and summarize}
hotel_summary <- 
  hotel_bookings %>%
  group_by(hotel) %>%
  summarise(average_lead_time=mean(lead_time),
            min_lead_time=min(lead_time),
            max_lead_time=max(lead_time))
```

Check out your new dataset using head() again:

```{r}
head(hotel_summary)
```

In [16]:
df.groupby('hotel').describe()

is_canceled                                               \
                   count      mean       std  min  25%  50%  75%  max   
hotel                                                                   
City Hotel       79330.0  0.417270  0.493111  0.0  0.0  0.0  1.0  1.0   
Resort Hotel     40060.0  0.277634  0.447837  0.0  0.0  0.0  1.0  1.0   

             lead_time              ... required_car_parking_spaces       \
                 count        mean  ...                         75%  max   
hotel                               ...                                    
City Hotel     79330.0  109.735724  ...                         0.0  3.0   
Resort Hotel   40060.0   92.675686  ...                         0.0  8.0   

             total_of_special_requests                                     \
                                 count      mean       std  min  25%  50%   
hotel                                                                       
City Hotel                     79330.0  0.546918  0.780776  0.0  0.0  0.0   
Resort Hotel                   40060.0  0.619770  0.813930  0.0  0.0  0.0   

                        
              75%  max  
hotel                   
City Hotel    1.0  5.0  
Resort Hotel  1.0  5.0  

[2 rows x 160 columns]

## Activity Wrap Up
Being able to manipulate data is a key skill for working in `R`. After this activity, you should be more familiar with functions that allow you to change your data, such as `arrange()`, `group_by()`, and `filter()`. You also have some experience using statistical summaries to make insights into your data. You can continue to practice these skills by modifying the code chunks in the rmd file, or use this code as a starting point in your own project console. As you practice, consider how performing tasks is similar and different in `R` compared to other tools you have learned throughout this program. 

